packer/linux/conf/bin/bk-install-elastic-stack.sh

```diff
@@ -37,16 +37,46 @@ PLUGINS_ENABLED=()
 [[ $DOCKER_LOGIN_PLUGIN_ENABLED == "true" ]] && PLUGINS_ENABLED+=("docker-login")
 
 # cfn-env is sourced by the environment hook in builds
-cat << EOF > /var/lib/buildkite-agent/cfn-env
-export DOCKER_VERSION=$DOCKER_VERSION
-export BUILDKITE_STACK_NAME=$BUILDKITE_STACK_NAME
-export BUILDKITE_STACK_VERSION=$BUILDKITE_STACK_VERSION
-export BUILDKITE_AGENTS_PER_INSTANCE=$BUILDKITE_AGENTS_PER_INSTANCE
-export BUILDKITE_SECRETS_BUCKET=$BUILDKITE_SECRETS_BUCKET
-export AWS_DEFAULT_REGION=$AWS_REGION
-export AWS_REGION=$AWS_REGION
-export PLUGINS_ENABLED="${PLUGINS_ENABLED[*]-}"
-export BUILDKITE_ECR_POLICY=${BUILDKITE_ECR_POLICY:-none}
+
+# We will create it in two steps so that we don't need to go crazy with quoting and escaping. The
+# first sets up a helper function, the second populates the default values for some environment
+# variables.
+
+# Step 1: Helper function.  Note that we clobber the target file and DO NOT apply variable
+# substitution, this is controlled by the double-quoted "EOF".
+cat <<- "EOF" > /var/lib/buildkite-agent/cfn-env
+	# The Buildkite agent sets a number of variables such as AWS_DEFAULT_REGION to fixed values which
+	# are determined at AMI-build-time.  However, sometimes a user might want to override such variables
+	# using an env: block in their pipeline.yml.  This little helper is sets the environment variables
+	# buildkite-agent and plugins expect, except if a user want to override them, for example to do a
+	# deployment to a region other than where the Buildkite agent lives.
+	function set_unless_present() {
+	    local target=$1
+	    local value=$2
+
+	    if [[ -v "${target}" ]]; then
+	        echo "^^^ +++"
+	        echo "⚠️ ${target} already set, NOT overriding! (current value: \"${!target}\")"
+	    else
+	        echo "export ${target}=\"${value}\""
+	        declare -gx "${target}=${value}"
+	    fi
+	}
+EOF
+
+# Step 2: Populate the default variable values.  This time, we append to the file, and allow
+# variable substitution.
+cat << EOF >> /var/lib/buildkite-agent/cfn-env
+
+set_unless_present "AWS_DEFAULT_REGION" "$AWS_REGION"
+set_unless_present "AWS_REGION" "$AWS_REGION"
+set_unless_present "BUILDKITE_AGENTS_PER_INSTANCE" "$BUILDKITE_AGENTS_PER_INSTANCE"
```

Comment:
I think we should only apply the `set_unless_present` function to the AWS environment variables above ☝️ 

These others are more a description of the stack you find yourself in rather than a configurable option that changes the behaviour of your job, e.g. setting `BUILDKITE_AGENTS_PER_INSTANCE` doesn’t spawn more agents, setting `DOCKER_VERSION` doesn’t route the Docker CLI commands to a second install etc

What do you think? 😄

Reply:
Yes!  Good point.  I was just mechanistically replacing-all, without thinking much about the role of the various variables.  I'll update as you suggest.

Reply:
Would you say the same for `PLUGINS_ENABLED` and `BUILDKITE_ECR_POLICY`, they're both more a statement of fact than a tweakable knob?